mrmustard/lab_dev/circuit_components.py

```diff
@@ -233,6 +233,42 @@ def __matmul__(self, other: CircuitComponent) -> CircuitComponent:
 
         return CircuitComponent.from_attributes("", wires_ret, representation_ret)
 
+    def __rshift__(self, other: CircuitComponent) -> CircuitComponent:
+        r"""
+        Contracts ``self`` and ``other`` as it would in a circuit, adding the adjoints when
+        they are missing.
+        """
+        if self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has ket and bra
+            return self @ other
+
+        if self.wires.ket and self.wires.bra and other.wires.ket and not other.wires.bra:
+            # self has ket and bra, other has ket
+            return self @ other @ other.adjoint
+
+        if self.wires.ket and self.wires.bra and not other.wires.ket and other.wires.bra:
+            # self has ket and bra, other has and bra
+            return self @ other @ other.adjoint
+
+        if self.wires.ket and not self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has ket, other has ket and bra
+            return self @ self.adjoint @ other
+
+        if not self.wires.ket and self.wires.bra and other.wires.ket and other.wires.bra:
+            # self has bra, other has ket and bra
+            return self @ self.adjoint @ other
```

Comment:
The same, maybe it can merge with 

`if (self.wires.ket or self.wires.bra) and other.wires.ket and other.wires.bra:`